EIPS/eip-7412.md

```diff
@@ -0,0 +1,167 @@
+
+With standards for both writing and reading cross-chain data, protocol developers will be able to create abstractions for asynchronicity (a topic thoroughly explored in other software engineering contexts). This will enable the development of highly sophisticated protocols that do not suffer from scaling constraints.
+
+[ERC-3668](./eip-3668.md) introduced the use of reverts for requiring off-chain data, but there are various challenges introduced by the specifics of that standard which are outlined in the _Rationale_ section below. By leveraging multicalls rather than callback functions, the standard proposed here is able to overcome some of these constraints.
+
+## Specification
+
+A contract implementing this standard MUST revert with the following error whenever off-chain data is required:
+
+```solidity
+error OracleDataRequired(address oracleContract, bytes oracleQuery)
+```
+
+`oracleQuery` specifies the off-chain data that is being required. Valid data formats for this parameter are specific to the oracle ID used. This might include chain id, contract address, function signature, payload, and timestamp/"latest" for cross-chain reads. For price feeds, it could include a ticker symbol and timestamp/"latest".
+
+`oracleContract` is the address of the contract which can verify the off-chain data and provide it to the contract to avoid the `OracleDataRequired` error. This contract MUST implement the following interface:
+
+```solidity
+interface IERC7412 {
+  function oracleId() view external returns (bytes32 oracleId);
+  function fulfillOracleData(bytes oracleQuery, bytes signedOffchainData) payable external;
+}
+```
+
+`oracleId` is a unique identifier that references the decentralized oracle network that generates the desired signed off-chain data. Oracle IDs would be analogous to Chain IDs in the Ethereum ecosystem. Clients are expected to resolve a gateway that corresponds to an Oracle ID, similar to how clients are expected to resolve an RPC endpoint based on a Chain ID.
+
+The contract implementing the `IOracleContract` interface MUST revert with the following error message if it requires payment to fulfill the oracle data query:
+
+```solidity
+error OracleFeeRequired(uint feeAmount)
+```
+
+`feeAmount` specifies the amount of native gas tokens required to execute the `fulfillOracleData` function, denominated in wei. This error MUST be resolved if the caller provides sufficient gas such that the fee amount can be collected by the oracle contract. The contract MAY NOT return gas tokens if they are provided in excess of the `feeAmount`. In practice, we would expect the fee amount to remain relatively stable, if not constant.
```

Comment:
How the fee is collected seems a bit vague, at least to me, a non-solidity expert. If it's intentionally left out, maybe include a sentence to that effect?